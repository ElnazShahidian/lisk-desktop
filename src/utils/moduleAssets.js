import { MODULE_ASSETS_NAME_ID_MAP, moduleAssetSchemas } from '@constants';
import { getSchemas } from './api/transaction';

const getModuleAssetSenderLabel = (t = str => str) => ({
  [MODULE_ASSETS_NAME_ID_MAP.transfer]: t('Sender'),
  [MODULE_ASSETS_NAME_ID_MAP.reclaimLSK]: t('Sender'),
  [MODULE_ASSETS_NAME_ID_MAP.unlockToken]: t('Sender'),
  [MODULE_ASSETS_NAME_ID_MAP.voteDelegate]: t('Voter'),
  [MODULE_ASSETS_NAME_ID_MAP.registerDelegate]: t('Account nickname'),
  [MODULE_ASSETS_NAME_ID_MAP.registerMultisignatureGroup]: t('Registrant'),
});

const getModuleAssetTitle = (t = str => str) => ({
  [MODULE_ASSETS_NAME_ID_MAP.transfer]: t('Send'),
  [MODULE_ASSETS_NAME_ID_MAP.reclaimLSK]: t('Reclaim'),
  [MODULE_ASSETS_NAME_ID_MAP.unlockToken]: t('Unlock'),
  [MODULE_ASSETS_NAME_ID_MAP.voteDelegate]: t('Vote'),
  [MODULE_ASSETS_NAME_ID_MAP.registerDelegate]: t('Register Delegate'),
  [MODULE_ASSETS_NAME_ID_MAP.registerMultisignatureGroup]: t('Register Multisignature Group'),
});

const retrieveSchemas = async ({ serviceUrl }) => {
  const response = await getSchemas({ baseUrl: serviceUrl });
  response.data.forEach((data) => {
    moduleAssetSchemas[data.moduleAssetId] = data.schema;
  });
};

const splitModuleAndAssetIds = (moduleAssetId) => {
  const [moduleID, assetID] = moduleAssetId.split(':');
  return [Number(moduleID), Number(assetID)];
};

const joinModuleAndAssetIds = ({ moduleID, assetID }) => [moduleID, assetID].join(':');

export {
  retrieveSchemas,
  getModuleAssetTitle,
  joinModuleAndAssetIds,
  splitModuleAndAssetIds,
  getModuleAssetSenderLabel,
};
