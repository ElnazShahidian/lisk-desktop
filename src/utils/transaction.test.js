import { MODULE_ASSETS_NAME_ID_MAP } from '@constants';
import { splitModuleAndAssetIds } from '@utils/moduleAssets';
import {
  getTxAmount,
  transformTransaction,
  containsTransactionType,
  createTransactionObject,
} from './transaction';
import accounts from '../../test/constants/accounts';

describe('API: LSK Transactions', () => {
  describe('getTxAmount', () => {
    it('should return amount of transfer in Beddows', () => {
      const tx = {
        moduleAssetId: MODULE_ASSETS_NAME_ID_MAP.transfer,
        asset: { amount: 100000000 },
      };

      expect(getTxAmount(tx)).toEqual(tx.asset.amount);
    });

    it('should return amount of votes in Beddows', () => {
      const tx = {
        title: MODULE_ASSETS_NAME_ID_MAP.voteDelegate,
        moduleAssetId: MODULE_ASSETS_NAME_ID_MAP.voteDelegate,
        asset: {
          votes: [
            {
              amount: '100000000',
            },
            {
              amount: '100000000',
            },
          ],
        },
      };

      expect(getTxAmount(tx)).toEqual(200000000);
    });

    it('should return amount of unlock in Beddows', () => {
      const tx = {
        title: MODULE_ASSETS_NAME_ID_MAP.unlockToken,
        moduleAssetId: MODULE_ASSETS_NAME_ID_MAP.unlockToken,
        asset: {
          unlockingObjects: [
            {
              amount: '100000000',
            },
            {
              amount: '100000000',
            },
          ],
        },
      };

      expect(getTxAmount(tx)).toEqual(200000000);
    });
  });

  describe('createTransactionObject', () => {
    it('creates a transaction object for transfer transaction', () => {
      const tx = {
        senderPublicKey: '',
        nonce: 1,
        recipient: '',
        amount: '1',
        fee: '1000000',
        data: 'test',
      };
      const txObj = createTransactionObject(
        tx,
        MODULE_ASSETS_NAME_ID_MAP.transfer,
      );

      expect(txObj).toMatchSnapshot();
    });

    it('creates a transaction object for vote transaction', () => {
      const tx = {
        senderPublicKey: '',
        nonce: 1,
        recipient: '',
        amount: '1',
        fee: '1000000',
        votes: [
          { amount: '100', delegateAddress: accounts.genesis.summary.address },
          {
            amount: '-100',
            delegateAddress: accounts.delegate.summary.address,
          },
        ],
      };
      const txObj = createTransactionObject(
        tx,
        MODULE_ASSETS_NAME_ID_MAP.voteDelegate,
      );

      expect(txObj).toMatchSnapshot();
    });

    it('creates a transaction object for delegate registration transaction', () => {
      const tx = {
        senderPublicKey: '',
        nonce: 1,
        fee: '1000000',
        username: 'username',
      };
      const txObj = createTransactionObject(
        tx,
        MODULE_ASSETS_NAME_ID_MAP.registerDelegate,
      );

      expect(txObj).toMatchSnapshot();
    });

    it('creates a transaction object for reclaimLSK transaction', () => {
      const tx = {
        senderPublicKey: '',
        nonce: 1,
        fee: '1000000',
        amount: '10000000',
      };
      const txObj = createTransactionObject(
        tx,
        MODULE_ASSETS_NAME_ID_MAP.reclaimLSK,
      );

      expect(txObj).toMatchSnapshot();
    });

    it('creates a transaction object for registerMultisignatureGroup transaction', () => {
      const tx = {
        senderPublicKey: '',
        nonce: 1,
        fee: '1000000',
        amount: '10000000',
        numberOfSignatures: 2,
        mandatoryKeys: [accounts.genesis.summary.publicKey, accounts.delegate.summary.publicKey],
        optionalKeys: [accounts.second_passphrase_account.summary.publicKey],
      };
      const txObj = createTransactionObject(
        tx,
        MODULE_ASSETS_NAME_ID_MAP.registerMultisignatureGroup,
      );

      expect(txObj).toMatchSnapshot();
    });
  });

  describe('transformTransaction', () => {
    const binaryAddress = 'd04699e57c4a3846c988f3c15306796f8eae5c1c';

    it('should a transfer transaction with type signature of lisk service', () => {
      const [moduleID, assetID] = splitModuleAndAssetIds(
        MODULE_ASSETS_NAME_ID_MAP.transfer,
      );
      const tx = {
        moduleID,
        assetID,
        fee: 0.1,
        nonce: 1,
        id: Buffer.from('123', 'hex'),
        senderPublicKey: accounts.genesis.summary.publicKey,
        asset: { amount: 100000000, recipientAddress: binaryAddress, data: '' },
      };

      expect(transformTransaction(tx)).toMatchSnapshot();
    });

    it('should a register delegate transaction with type signature of lisk service', () => {
      const [moduleID, assetID] = splitModuleAndAssetIds(
        MODULE_ASSETS_NAME_ID_MAP.registerDelegate,
      );
      const tx = {
        moduleID,
        assetID,
        fee: 0.1,
        nonce: 1,
        id: Buffer.from('123', 'hex'),
        senderPublicKey: accounts.genesis.summary.publicKey,
        asset: { username: 'super_delegate' },
      };

      expect(transformTransaction(tx)).toMatchSnapshot();
    });

    it('should a vote delegate transaction with type signature of lisk service', () => {
      const [moduleID, assetID] = splitModuleAndAssetIds(
        MODULE_ASSETS_NAME_ID_MAP.voteDelegate,
      );
      const tx = {
        moduleID,
        assetID,
        fee: 0.1,
        nonce: 1,
        id: Buffer.from('123', 'hex'),
        senderPublicKey: accounts.genesis.summary.publicKey,
        asset: {
          votes: [
            {
              amount: '100',
              delegateAddress: '123',
            },
          ],
        },
      };

      expect(transformTransaction(tx)).toMatchSnapshot();
    });

    it('should transform a reclaimLSK transaction', () => {
      const [moduleID, assetID] = splitModuleAndAssetIds(
        MODULE_ASSETS_NAME_ID_MAP.reclaimLSK,
      );
      const tx = {
        moduleID,
        assetID,
        fee: 0.1,
        nonce: 1,
        id: Buffer.from('123', 'hex'),
        senderPublicKey: accounts.genesis.summary.publicKey,
        asset: {
          amount: '100',
        },
      };

      expect(transformTransaction(tx)).toMatchSnapshot();
    });

    it('should transform a registerMultisignatureGroup transaction', () => {
      const [moduleID, assetID] = splitModuleAndAssetIds(
        MODULE_ASSETS_NAME_ID_MAP.registerMultisignatureGroup,
      );
      const mandatoryKeys = [accounts.genesis.summary.publicKey, accounts.delegate.summary.publicKey].map(key => Buffer.from(key, 'hex'));
      const optionalKeys = [accounts.second_passphrase_account.summary.publicKey].map(key => Buffer.from(key, 'hex'));

      const tx = {
        moduleID,
        assetID,
        id: Buffer.from('123', 'hex'),
        senderPublicKey: accounts.genesis.summary.publicKey,
        nonce: 1,
        fee: '1000000',
        amount: '10000000',
        asset: {
          numberOfSignatures: 2,
          mandatoryKeys,
          optionalKeys,
        },
      };

      expect(transformTransaction(tx)).toMatchSnapshot();
    });
  });

  describe('containsTransactionType', () => {
    const { transfer, voteDelegate } = MODULE_ASSETS_NAME_ID_MAP;

    it('should return true', () => {
      let pending = [{ moduleAssetId: voteDelegate }];
      expect(containsTransactionType(pending, voteDelegate)).toEqual(true);

      pending = [{ moduleAssetId: transfer }, { moduleAssetId: voteDelegate }];
      expect(containsTransactionType(pending, voteDelegate)).toEqual(true);
    });

    it('should return false', () => {
      let pending = [];
      expect(containsTransactionType(pending, voteDelegate)).toEqual(false);

      pending = [{ moduleAssetId: transfer }];
      expect(containsTransactionType(pending, voteDelegate)).toEqual(false);
    });
  });
});
