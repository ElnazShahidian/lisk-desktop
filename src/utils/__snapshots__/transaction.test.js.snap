// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`API: LSK Transactions createTransactionObject creates a transaction object for delegate registration transaction 1`] = `
Object {
  "asset": Object {
    "username": "username",
  },
  "assetID": 0,
  "fee": 1000000n,
  "moduleID": 5,
  "nonce": 1n,
  "senderPublicKey": Object {
    "data": Array [],
    "type": "Buffer",
  },
  "signatures": Array [],
}
`;

exports[`API: LSK Transactions createTransactionObject creates a transaction object for reclaimLSK transaction 1`] = `
Object {
  "asset": Object {
    "amount": 10000000n,
  },
  "assetID": 0,
  "fee": 1000000n,
  "moduleID": 1000,
  "nonce": 1n,
  "senderPublicKey": Object {
    "data": Array [],
    "type": "Buffer",
  },
  "signatures": Array [],
}
`;

exports[`API: LSK Transactions createTransactionObject creates a transaction object for registerMultisignatureGroup transaction 1`] = `
Object {
  "asset": Object {
    "mandatoryKeys": Array [
      Object {
        "data": Array [
          15,
          233,
          163,
          241,
          162,
          27,
          85,
          48,
          242,
          127,
          135,
          164,
          20,
          181,
          73,
          231,
          154,
          148,
          11,
          242,
          79,
          223,
          43,
          47,
          5,
          231,
          242,
          42,
          238,
          236,
          200,
          106,
        ],
        "type": "Buffer",
      },
      Object {
        "data": Array [
          134,
          73,
          152,
          121,
          68,
          141,
          27,
          2,
          21,
          213,
          156,
          191,
          7,
          136,
          54,
          227,
          215,
          217,
          210,
          120,
          45,
          86,
          162,
          39,
          74,
          86,
          135,
          97,
          191,
          243,
          111,
          25,
        ],
        "type": "Buffer",
      },
    ],
    "numberOfSignatures": 2,
    "optionalKeys": Array [
      Object {
        "data": Array [
          250,
          185,
          210,
          97,
          234,
          5,
          11,
          158,
          50,
          109,
          126,
          17,
          88,
          126,
          204,
          195,
          67,
          162,
          14,
          100,
          226,
          157,
          135,
          129,
          181,
          15,
          208,
          102,
          131,
          202,
          204,
          136,
        ],
        "type": "Buffer",
      },
    ],
  },
  "assetID": 0,
  "fee": 1000000n,
  "moduleID": 4,
  "nonce": 1n,
  "senderPublicKey": Object {
    "data": Array [],
    "type": "Buffer",
  },
  "signatures": Array [],
}
`;

exports[`API: LSK Transactions createTransactionObject creates a transaction object for transfer transaction 1`] = `
Object {
  "asset": Object {
    "amount": 1n,
    "data": "test",
    "recipientAddress": Object {
      "data": Array [],
      "type": "Buffer",
    },
  },
  "assetID": 0,
  "fee": 1000000n,
  "moduleID": 2,
  "nonce": 1n,
  "senderPublicKey": Object {
    "data": Array [],
    "type": "Buffer",
  },
  "signatures": Array [],
}
`;

exports[`API: LSK Transactions createTransactionObject creates a transaction object for vote transaction 1`] = `
Object {
  "asset": Object {
    "votes": Array [
      Object {
        "amount": 100n,
        "delegateAddress": Object {
          "data": Array [
            208,
            70,
            153,
            229,
            124,
            74,
            56,
            70,
            201,
            136,
            243,
            193,
            83,
            6,
            121,
            111,
            142,
            174,
            92,
            28,
          ],
          "type": "Buffer",
        },
      },
      Object {
        "amount": -100n,
        "delegateAddress": Object {
          "data": Array [
            183,
            120,
            252,
            149,
            185,
            240,
            116,
            7,
            230,
            64,
            157,
            115,
            200,
            175,
            25,
            25,
            217,
            3,
            80,
            2,
          ],
          "type": "Buffer",
        },
      },
    ],
  },
  "assetID": 1,
  "fee": 1000000n,
  "moduleID": 5,
  "nonce": 1n,
  "senderPublicKey": Object {
    "data": Array [],
    "type": "Buffer",
  },
  "signatures": Array [],
}
`;

exports[`API: LSK Transactions transformTransaction should a register delegate transaction with type signature of lisk service 1`] = `
Object {
  "asset": Object {
    "username": "super_delegate",
  },
  "fee": "0.1",
  "id": "12",
  "moduleAssetId": "5:0",
  "nonce": "1",
  "sender": Object {
    "address": "lskdxc4ta5j43jp9ro3f8zqbxta9fn6jwzjucw7yt",
    "publicKey": "0fe9a3f1a21b5530f27f87a414b549e79a940bf24fdf2b2f05e7f22aeeecc86a",
  },
  "signatures": undefined,
}
`;

exports[`API: LSK Transactions transformTransaction should a transfer transaction with type signature of lisk service 1`] = `
Object {
  "asset": Object {
    "amount": "100000000",
    "data": "",
    "recipient": Object {
      "address": "lskzzzz3xu4xpzz6x2zzuzzbvzpz2zzrvz3zzxuzz3mzozzox24z2zzuzzzzzvuzz3z577dz7",
    },
  },
  "fee": "0.1",
  "id": "12",
  "moduleAssetId": "2:0",
  "nonce": "1",
  "sender": Object {
    "address": "lskdxc4ta5j43jp9ro3f8zqbxta9fn6jwzjucw7yt",
    "publicKey": "0fe9a3f1a21b5530f27f87a414b549e79a940bf24fdf2b2f05e7f22aeeecc86a",
  },
  "signatures": undefined,
}
`;

exports[`API: LSK Transactions transformTransaction should a vote delegate transaction with type signature of lisk service 1`] = `
Object {
  "asset": Object {
    "votes": Array [
      Object {
        "amount": 100,
        "delegateAddress": "lskzpxzckpryh",
      },
    ],
  },
  "fee": "0.1",
  "id": "12",
  "moduleAssetId": "5:1",
  "nonce": "1",
  "sender": Object {
    "address": "lskdxc4ta5j43jp9ro3f8zqbxta9fn6jwzjucw7yt",
    "publicKey": "0fe9a3f1a21b5530f27f87a414b549e79a940bf24fdf2b2f05e7f22aeeecc86a",
  },
  "signatures": undefined,
}
`;

exports[`API: LSK Transactions transformTransaction should transform a reclaimLSK transaction 1`] = `
Object {
  "asset": Object {
    "amount": "100",
  },
  "fee": "0.1",
  "id": "12",
  "moduleAssetId": "1000:0",
  "nonce": "1",
  "sender": Object {
    "address": "lskdxc4ta5j43jp9ro3f8zqbxta9fn6jwzjucw7yt",
    "publicKey": "0fe9a3f1a21b5530f27f87a414b549e79a940bf24fdf2b2f05e7f22aeeecc86a",
  },
  "signatures": undefined,
}
`;

exports[`API: LSK Transactions transformTransaction should transform a registerMultisignatureGroup transaction 1`] = `
Object {
  "asset": Object {
    "mandatoryKeys": Array [
      "0fe9a3f1a21b5530f27f87a414b549e79a940bf24fdf2b2f05e7f22aeeecc86a",
      "86499879448d1b0215d59cbf078836e3d7d9d2782d56a2274a568761bff36f19",
    ],
    "numberOfSignatures": 2,
    "optionalKeys": Array [
      "fab9d261ea050b9e326d7e11587eccc343a20e64e29d8781b50fd06683cacc88",
    ],
  },
  "fee": "1000000",
  "id": "12",
  "moduleAssetId": "4:0",
  "nonce": "1",
  "sender": Object {
    "address": "lskdxc4ta5j43jp9ro3f8zqbxta9fn6jwzjucw7yt",
    "publicKey": "0fe9a3f1a21b5530f27f87a414b549e79a940bf24fdf2b2f05e7f22aeeecc86a",
  },
  "signatures": undefined,
}
`;
